### ISSUE TITLE
Handle HTTPS urls properly

### ISSUE THREAD

**Issue by dsadinoff** [leejo/AWS-S3#7]
In order to properly support TLS  urls, the API will need to properly handle the correct S3 endpoints for a bucket. 
e.g. 

http://bucket-name.s3.amazonaws.com/key
wants to be  e.g.
https://s3-eu-west-1.amazonaws.com/bucketname/key

the signature is not affected by this transformation, but the S3::Bucket code really ought to do the right thing here.

I may try to implement.

**Comment by leejo** [comment]
👍  - i recall something about this previously, maybe in an RT ticket before i took over maint of the module.

**Comment by dsadinoff** [comment]
so the Bucket code wants to do a  GET Bucket location...
http://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketGETlocation.html

...and then look that location up in this table:
http://docs.aws.amazon.com/general/latest/gr/rande.html#s3_region

...to get the proper endpoint.

**Comment by dsadinoff** [comment]
Might as well throw in a dualStack option while we're in there.

So the question is what should the API be?
right now we've got
AWS::File::signed_url( $expiration? ) 

options include:
AWS::File::signed_url( $expiration?, $tls? , $dualStack? ) 

or perhaps options style:

AWS::File::signed_url( expiration => $expiration, tls => $tls, dualStack =>  $dualStack ) ;
...with fallback to the old style if passed in one argument.

...or we can add a different method entirely.

**Comment by leejo** [comment]
> `AWS::File::signed_url( expiration => $expiration, tls => $tls, dualStack => $dualStack ) ;`

This is the most flexible i think.

**Comment by simbabque** [comment]
@dsadinoff did you start implementing this? I'm looking for something to do for the Pull Request Challenge.

**Comment by dsadinoff** [comment]
Nope,  It's in the "maybe someday" pile, with a workaround like so:

```perl
method getSignedUrl( $bucketName, $key, $expiry?){
    my $s3 = $self->s3;
    my $bucket = $s3->bucket($bucketName );
    my $file = $bucket->file($key);

    return unless $file;
    my $urlBad = $file->signed_url(  $expiry );
    (my $urlTmp = $urlBad) =~s{%2F}{/}g; # https://github.com/leejo/AWS-S3/issues/6
    (my $urlTmp2 = $urlTmp) =~s{%3A}{:}g; # https://github.com/leejo/AWS-S3/issues/6
    (my $url = $urlTmp2) =~s{%3D}{=}g; # https://github.com/leejo/AWS-S3/issues/6

    my $endpoint = 's3-eu-west-1.amazonaws.com'; # FIX derive from bucket
    # https://github.com/leejo/AWS-S3/issues/7
    $url =~ s{http://$bucketName\.s3.amazonaws.com/}{https://$endpoint/$bucketName/};

    return $url;
}
```

**Comment by dsadinoff** [comment]
apologies, I'm a big talker :)

**Comment by dsadinoff** [comment]
Have at it!

note, again, the proper implementation API probably wants to do a 
http://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketGETlocation.html
in order to determine the correct endpoint, and maybe someday do some caching of that value per bucket.

**Comment by simbabque** [comment]
@dsadinoff I think you posted your code into the wrong ticket. :)

**Comment by leejo** [comment]
I've just pushed v1.00 to CPAN, which should resolve this issue.
